src/apps/vguest/vguest_dev.lua

```diff
@@ -0,0 +1,227 @@
+
+module(..., package.seeall)
+
+local ffi       = require("ffi")
+local C         = ffi.C
+local S         = require('syscall')
+local pci       = require("lib.hardware.pci")
+local bit       = require('bit')
+local lib       = require("core.lib")
+local link      = require("core.link")
+local packet    = require("core.packet")
+local checksum  = require("lib.checksum")
+local gvring    = require('apps.vguest.guest_vring')
+require('lib.virtio.virtio_h')
+
+
+local function fstruct(def)
+   local struct = {}
+   local offset = 0
+   for ct, fld in def:gmatch('([%a_][%w_]*)%s+([%a_][%w_]*);') do
+      ct = ffi.typeof(ct)
+      struct[fld] = {
+         fieldname = fld,
+         ct = ct,
+         size = ffi.sizeof(ct),
+         offset = offset,
+      }
+      offset = offset + struct[fld].size
+   end
+   return struct, offset
+end
+
+local function fieldrd(field, fd)
+   local buf = ffi.typeof('$ [1]', field.ct)()
+   local r, err = fd:pread(buf, field.size, field.offset)
+   if not r then error(err) end
+   return buf[0]
+end
+
+local function fieldwr(field, fd, val)
+--    print ('fieldwr', field.ct, field.size, field.offset)
+   local buf = ffi.typeof('$ [1]', field.ct)()
+   buf[0] = val
+--    print ('val:', val, buf, buf[0])
+   assert(fd:seek(field.offset))
+--    print ('tell:', fd:tell())
+   local r, err = fd:write(buf, field.size)
+   if not r then error(err) end
+   return buf[0]
+end
+
+local function openBar(fname, struct)
+   local fd, err = S.open(fname, 'rdwr')
+   if not fd then error(err) end
+   return setmetatable ({
+      fd = fd,
+      struct = struct,
+      close = function(self) return self.fd:close() end,
+   }, {
+      __index = function (self, key)
+         return fieldrd(self.struct[key], self.fd)
+      end,
+      __newindex = function (self, key, value)
+         return fieldwr(self.struct[key], self.fd, value)
+      end,
+   })
+end
+
+virtio_pci_bar0 = fstruct[[
+   uint32_t host_features;
+   uint32_t guest_features;
+   uint32_t queue_pfn;
+   uint16_t queue_num;
+   uint16_t queue_sel;
+   uint16_t queue_notify;
+   uint8_t status;
+   uint8_t isr;
+   uint16_t config_vector;
+   uint16_t queue_vector;
+]]
+
+
+VGdev = {}
+VGdev.__index = VGdev
+
+
+function VGdev:new(args)
+   local min_features = 0 -- C.VIRTIO_F_VERSION_1?
+   local want_features = C.VIRTIO_NET_F_CSUM
+                        + C.VIRTIO_NET_F_MAC
+--                         + C.VIRTIO_F_VERSION_1
+   pci.unbind_device_from_linux (args.pciaddr)
+
+   local bar = openBar(pci.path(args.pciaddr..'/resource0'), virtio_pci_bar0)
+--    for k,v in pairs(virtio_pci_bar0) do
+--       print (string.format('%s: %X', v.fieldname, bar[v.fieldname]))
+--    end
+
+   bar.status = 0           -- reset device
+   bar.status = bit.bor(bar.status, 1)           -- acknowledge
+   -- check something
+   bar.status = bit.bor(bar.status, 2)           -- driver
+   local features = bar.host_features
+   print ('host_features', features)
```

Comment:
It seems this print messages are meant to be debug information. I'd put them behind a debug flag.